### ISSUE TITLE
vmap breaks gradient on custom c++ functions

### ISSUE THREAD

**Issue by vmoens** [pytorch/pytorch#102720]
### 🐛 Describe the bug

When calling vmap with custom c++ functions, the gradients are broken:
I'm basing this on the [safetanh from torchrl](https://github.com/pytorch/rl/blob/fa45472da29cc1d07b3ae57bdfe094207deeff13/torchrl/csrc/utils.h#L13-L31)

```python
>>> import torch
>>> from torch import vmap
>>> from torchrl._torchrl import safeatanh, safetanh
>>> safetanh(torch.zeros(3, requires_grad=True), 1e-6)
tensor([0., 0., 0.], grad_fn=<CppNode<SafeTanh>>)
>>> vmap(safetanh, (0, None))(torch.zeros(3, requires_grad=True), 1e-6)
tensor([0., 0., 0.])
>>> # for ref
>>> vmap(torch.tanh)(torch.zeros(3, requires_grad=True))
tensor([0., 0., 0.], grad_fn=<TanhBackward0>)
```

I guess I should fix the c++ code in some way, but maybe we should raise an exception somewhere here?

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @Chillee @samdow @kshitij12345 @janeyx99 

### Versions

2.1.0.dev20230525

## Edit

I followed https://pytorch.org/docs/master/notes/extending.func.html and I managed to make it work in python.
The python error was very explicit, that really helped.

**Comment by zou3519** [comment]
Hi-pri for silent incorrectness

**Comment by albanD** [comment]
@zou3519 will the new custom op API allow users to define these better?

**Comment by zou3519** [comment]
> will the new custom op API allow users to define these better?

New custom op API is in Python, but yeah you would be able to define a batching rule with it. Note that you can already define a batching rule with (python) autograd.Function.

**Comment by janeyx99** [comment]
Following group discussion: this can be fixed by raising an error similar to the Python custom functions. ccing @zou3519 for more context
